src/crystal/scheduler.cr

```diff
@@ -0,0 +1,86 @@
+require "./event_loop"
+require "fiber"
+require "thread"
+
+# :nodoc:
+#
+# Schedulers are tied to a thread, and must only ever be accessed from within
+# this thread.
+#
+# Only the class methods are public and safe to use. Instance methods are
+# protected and must never be called directly.
+class Crystal::Scheduler
+  def self.current_fiber : Fiber
+    Thread.current.scheduler.@current
+  end
+
+  def self.enqueue(fiber : Fiber) : Nil
+    Thread.current.scheduler.enqueue(fiber)
+  end
+
+  def self.enqueue(fibers : Enumerable(Fiber)) : Nil
+    Thread.current.scheduler.enqueue(fibers)
+  end
+
+  def self.reschedule : Nil
+    Thread.current.scheduler.reschedule
+  end
+
+  def self.resume(fiber : Fiber) : Nil
+    Thread.current.scheduler.resume(fiber)
+  end
+
+  def self.sleep(time : Time::Span) : Nil
+    Thread.current.scheduler.sleep(time)
+  end
+
+  def self.yield : Nil
+    Thread.current.scheduler.yield
+  end
+
+  def self.yield(fiber : Fiber) : Nil
+    Thread.current.scheduler.yield(fiber)
+  end
+
+  # :nodoc:
+  def initialize(@main : Fiber)
+    @current = @main
+    @runnables = Deque(Fiber).new
+  end
+
+  protected def enqueue(fiber : Fiber) : Nil
+    @runnables << fiber
+  end
+
+  protected def enqueue(fibers : Enumerable(Fiber)) : Nil
+    @runnables.concat fibers
+  end
+
+  protected def resume(fiber : Fiber) : Nil
+    current, @current = @current, fiber
+    GC.stack_bottom = fiber.@stack_bottom
+    Fiber.swapcontext(pointerof(current.@stack_top), fiber.@stack_top)
```

Comment:
Before the PR the swapcontext use `pointerof(fiber.@stack_top)`  why is that not needed? I see no changes in the assembler code in `x86_64` for example.

Reply:
Because the assembly needs a pointer to the `stack_top` pointer of the current fiber to write the current stack pointer to (i.e. `void**`), but we don't need to dereference the pointer to load the new `stack_top` pointer, and can just pass the pointer itself (i.e. `void*`).

AArch64 used that form, and I harmonized the calls.

Reply:
The difference is the `movq ($1), %rsp` vs `movq $1, %rsp` in the asm

Reply:
Ok, thanks for clarifying